### DISCUSSION TITLE
Create a "Link/Button" field

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#9062] [Ideas] [answered]
> Created by: Ryan (rswhite@gmail.com) on 2015/06/27 22:53:39 +0000
> Votes at time of UserVoice import: 3

-----------------------------------

Sometimes on a page it's necessary to give the client the option to add a button and/or customize button text, url, if the url should pull from an existing entry, if url should open in new tab, add a button class via a dropdown, etc.

Right now, you can easily do this with a Matrix block and restrict it to a max 1 block, but then the user has to "add the block" and it seems like an unnecessary step. Also, if you're needing to add a customizable link/button inside a matrix field, you can't add ANOTHER matrix field, so you'd have to create all those fields individually. It would be cool to see some way of grouping all of these things together into one field and just choose all the details you want associated with the link/button.

**Comment by angrybrad** [comment]
> Posted by Judd Lyon (jl@juddlyon.com) on 2016/01/28 20:46:03 +0000
-----------------------------------

This is a requirement on virtually every CMS-powered site. Linkit is great but not yet 2.5 compatible. It would be ideal not to rely on a third party add-on and have this be native functionality.

**Comment by angrybrad** [comment]
> Posted by Kaspar (bonjour@kasparallenbach.ch) on 2015/03/11 14:14:49 +0000
-----------------------------------

The LinkeIt plugin is your friend: https://github.com/fruitstudios/LinkIt

**Comment by brandonkelly** [comment]
I could see us adding a generic Link field once #1662 is implemented.

**Comment by jsunsawyer** [comment]
Is this still something you're considering?

**Comment by brandonkelly** [comment]
@jsunsawyer it’s a possibility down the road, but definitely not a priority for us right now.

**Comment by bfopma** [comment]
@brandonkelly we've used LinkIt heavily in our projects; with recent conversation in Discord about LinkIt ceasing to be under active development, is there opportunity to revisit this as a built-in fieldtype in Craft?

I'm not sure if this pain point is exclusive to us, but we've both relied heavily on LinkIt and have been burned by it during Craft upgrades. A first party link fieldtype would be a very welcome addition. Thank you for considering.

**Comment by Mosnar** [comment]
I wanted to poke this discussion since we're coming up on the Craft 4 release and there's no longer a stable release available for a "Link" field. This is definitely the biggest roadblock that will be holding us back from upgrading to Craft 4 (and was also the biggest roadblock holding us back from upgrading to Craft 3).

My opinion is that a native link field should be a high priority given the crucial role it plays in every website and the lack of a well-maintained third-party alternative.

**Reply by brandonkelly** [reply]
I think it’s a little early to consider the existing plugins abandoned. Half of our own first party plugins haven’t even been ported yet.

In any case, the earliest we’d consider doing this ourselves would be after it’s possible to query across multiple element types simultaneously.

**Reply by pixelmachine** [reply]
I think it's worth mentioning a year and a half after the last response @brandonkelly that both Linkit and Link field are both back in semi-abandoned states. Neither has seen even a bugfix update in nearly a year and bug reports and PRs go mostly unresolved.

They both work (with caveats) and maybe they'll get an update for Craft 5 at some point but a first party solution would be very welcome at this point.

I'm not sure about others but we use a lot of link fields on our projects and it's a shame to have an unreliable foundation for a fairly common function. It looks like the two plugins have an install base of around 43k people so there's definitely demand there.

**Reply by bfopma** [reply]
A huge +1 on this one. This is a significant concern for us as we look ahead to Craft 5. If the developers of Typed Link (or LinkIt) decide not to update their respective plugins we are going to have a very difficult upgrade path. 

Is there a possibility for Craft to formally adopt one of these, similar to the Feed Me acquisition?

**Reply by piotrpog** [reply]
@bfopma would be great if this plugin was adopted.

**Reply by brandonkelly** [reply]
It’s definitely on the nice-to-have list, but I can’t promise we’ll have a native link field in v5.

You can check out Verbb’s [Hyper](https://verbb.io/craft-plugins/hyper/features) plugin, which is quite good, and provides a migration path for [Typed Link Field](https://verbb.io/craft-plugins/hyper/docs/guides/migrating-from-typed-link), [LinkIt](https://verbb.io/craft-plugins/hyper/docs/guides/migrating-from-linkit), and [Link](https://verbb.io/craft-plugins/hyper/docs/guides/migrating-from-link).

**Reply by leevigraham** [reply]
+1 for https://verbb.io/craft-plugins/hyper/features

**Comment by cherrykoda** [comment]
I was about to make a new issue, but found this one. So I'll paste my details here instead.

A large portion of the community was relying on https://plugins.craftcms.com/typedlinkfield which is no longer being maintained. The remaining plugins available are 3rd party and cost money, but customers expect a link field that can handle entry/url/email/tel as a native CMS field. Other CMS platforms that compete with Craft CMS offer it, and its a standard in familiar UIs across apps, platforms etc.

**Minimum Viable Solution:**
A link field that handles choosing an entry, a URL, and option to open in new window

**Best Solution:**
MVS in addition to:
limiting the entry sections available to choose from
Assets Picker

**"Extra":**
noref/nofollow option on the link
aria support
Users Picker

**Reply by cherrykoda** [reply]
Please do note that this is something that makes Craft hard to sell, with the (justified!) price increases and no more support for the free plugin... conversations with existing clients to pay for another plugin have them going "what is this, WordPress?! I thought Craft was better than this!" 

So while trivial sounding, this would do a lot of damage control and patch up some authorship + plugin complaints. Thanks for the consideration.

**Reply by bfopma** [reply]
![image](https://github.com/craftcms/cms/assets/6729531/85009f8c-016d-423a-9ee6-ff1de5059353)

Through several major Craft version upgrades we've been recommended to use LinkIt (abandoned prior to Craft 3 with no migration option at launch), Typed Link (see above message that is now appearing in the CP), and now Hyper...

Love the Verbb plugins--we already rely on several for each project. But that's a lot of eggs in one basket. Relying on third-party developers for what arguably should be a core field type has proven to be a challenging strategy.

+100 to please consider adding a first-party link type

**Reply by davidhellmann** [reply]
Also my +1 for a core link field like hyper/linkit/typedlinkfield

**Reply by johndwells** [reply]
@cherrykoda I 100% agree with you on all points. Just to drive home how Craft stands starkly alone in it's lack of 1st party link field support:

[Link module for Drupal](https://www.drupal.org/project/link) now part of Drupal 8 core
[Link Fieldtype for Statamic](https://statamic.dev/fieldtypes/link)
~~[URL Fieldtype for ExpressionEngine](https://docs.expressionengine.com/v6/fieldtypes/url.html)~~ (nvmd, that's just a URL)
[Link field for Kirby](https://getkirby.com/docs/reference/panel/fields/link)
[URL type for Sanity.io](https://www.sanity.io/docs/url-type)
[ACF Link field type for Wordpress](https://www.advancedcustomfields.com/resources/link/) _Technicallay ACF is not core but it's defacto core for anyone who's done any WP development_

And also, Hyper is _such_ a good plugin that there's plenty of space for a "basic" link field in core while leaving plenty of market space for Hyper.

Craft 5 is the time guys. Seriously.

**Reply by johndwells** [reply]
I'm also chuckling/crying at the irony that one of the features we sell to clients about Craft is its ability to create deep relationships (aka "links") between content across their site... only to admit that what Craft can't do particularly well out of the box is create links to _anything else_. 🤦

**Reply by cherrykoda** [reply]
@johndwells we use the lack of reliance on plugins / not having an ecosystem where sites break or have major vulnerabilities due to too many plugins as a selling point too... this one was at least maskable by a free one but now it's become a technical debt / reckoning point with end users.

**Reply by mfell** [reply]
Yes ... +1K for a native Link Field within the core of Craft 5!

Typed Link Field is not supported and yes, there is an option to user another 3rd party plugin or build something with Matrix.
But I don't want to stick in another dependency to a 3rd party plugin and it's frustating to build workarounds and maybe a native Link Field comes along with Craft 5.x.

Now, it's upgrade time to Craft 5 (I love it!), but please give us this feature now.
By the way: TYPO3 (https://typo3.org/) is my other CMS in use -> Link Field is part of the core for decade.

**Reply by CreateSean** [reply]
Yes please. Also would be great if an option to migrate to the new plugin were created for sites on Craft 4.  Otherwise those sites are going to be unable to upgrade to Craft 5

**Reply by CreateSean** [reply]
Another thought - would it be possible for Pixel and Tonic to acquire the plugin and bake it into Craft?

**Reply by mihob** [reply]
A native field type would be great! This is a function that I need in virtually every project

**Reply by alexanderloewe** [reply]
We need a native link field @brandonkelly!

**Reply by stephanedemotte** [reply]
Typedlinkfield was one of the must popular plugin, please @brandonkelly listen the community and make the field exist in Craft 5 :)

**Reply by alexanderloewe** [reply]
Yes, do it. Do it ;)

**Reply by Mankeldor** [reply]
@brandonkelly:

As @cherrykoda mentioned: One of the key selling points to clients is a CMS which doesn't rely on 3rd party plugins. Especially not on paid plugins for a **core functionality**.

It would be great if this core functionality will be implemented. A switch from nice-to-have-list to need-to-be-implemented-list is long overdue (imho).

**Reply by stephanedemotte** [reply]
Look like it's really easy to re-create a simple link component with the new `Entry Type` system and `conditional` on Craft 5

<img width="1296" alt="Screenshot 2024-03-18 at 11 57 29" src="https://github.com/craftcms/cms/assets/2158130/4cacdfa9-c931-4168-bd10-aa0c282f8a47">

**Reply by mfell** [reply]
Ok ... is a link to an asset (e.g. pdf) or category possible?

**Reply by handplant** [reply]
Yes, you can extend the dropdown link type and then provide an AssetField via conditions.

**Reply by Mankeldor** [reply]
@stephanedemotte : That sounds promising. Although there will be "a lot of twig ifs" in templates, this is a practicable solution. Thank you.

Nonetheless a real link field would be much better.

**Reply by davidhellmann** [reply]
Conditionals are suitable, but you can also use entry types. It's a cleaner approach. You can do cleaner twig comps with render. I also started with the conditional approach, but it's much harder to maintain long-term. If you have to change something or need an additional field, you need to think about too much stuff. An entry type feels more robust to me.

**Reply by splendidrob** [reply]
@stephanedemotte How would you go about using this within another entry type - in the way that the link field can be used inside say a banner with text and a call to action? Or do you just have to create it every time you want to use it in separate matrix blocks for instance?

**Reply by thupsi** [reply]
@splendidrob You _could_ use a matrix with 1 min / 1 max blocks and use this like a reusable "field group", though this [could be made a little better](https://github.com/craftcms/cms/discussions/14015).

**Reply by cherrykoda** [reply]
This all seems like a lot of effort to work around not having a native link field. I know as devs we can solve for things but at what point is all the extra work just silly?

**Reply by RyanRoberts** [reply]
Typed Link Field has 32,581 active installs, I'd say that's a huge vote for a devs wishing Craft had a native link field 😆

**Reply by mfell** [reply]
Maybe a new approach:
Couldn't someone else adopt this plugin and bring it to Craft 5?
Unfortunately I'm "only" a frontend, I lack the backend knowledge. And I can't estimate whether it's really that much effort.

Regardless of this, it remains problematic that a plugin is required for this topic and the Craft CMS core does not provide this natively. (Just my 2 cents...)

**Reply by piotrpog** [reply]
We already have Hyper plugin which also provides same functionality and is very well made. The problem is that something so basic and necessary for content modeling is not first party, like for example ckeditor plugin.

**Reply by Mankeldor** [reply]
@piotrpog: Hyper plugin is not a solution to this case/problem. As @cherrykoda already mentioned above: We cannot justify to our clients to pay for a third party plugin to gain a core functionality. Especially that craft licence is now at 399$.

**Reply by piotrpog** [reply]
@Mankeldor 
Agreed, but i was just commenting on the issue that typed link field is no longer maintained and argued that for the time being hyper can be placeholder for that.
But we definitely need first party link plugin (or not even a plugin, just including that into craft core).

**Reply by mihob** [reply]
I'm with @Mankeldor  here. Hyper is a great plugin, but honestly, this is a basic functionality that shouldn't generate additional costs. Not to mention the effort involved in migrating extensive existing content to a new solution.

**Reply by mfell** [reply]
@piotrpog Yes, the Hyper Plugin is very well done ... no doubt. But I don't want to stick in a 3rd party dependency again. And I do use plugins, of course, but in this case, we talk about a core function and that is the frustrating aspect for me. 399$ for a Craft license is expensive, but Craft is it worth ... I'm fine with this. But I think, we can expect this link function within the Craft core.

**Reply by piotrpog** [reply]
@mfell @mihob I think the worst offender here is seo functionality actually. I can sometimes live without link plugin, but seo needs to be included EVERYWHERE. 
And now every site needs additional 99$ for seomatic. Which is excellent plugin but for some sites that just require basic stuff it is overkill.
Other free seo plugins are just not up to par with it. I sometimes use seomate which works well, but it completely lacks admin interface, so i need to set up some fields manually which is just cumbersome.
If we are talking first party plugins, i would love to see some first party seo solution, something simple that handles just description, title, robots, open graph and sitemap. Even before we get first party link plugin.

**Reply by mfell** [reply]
@piotrpog That's another discussion ... we should not hijack this thread.

**Reply by alexanderloewe** [reply]
This is how I see it:

Craft CMS now costs $399, but hasn't gained much in terms of actual features. Most things are still the same as they were many years ago. The new matrix field does not really fulfill the requirements according to my inquiries with colleagues, so many developers continue to use Neo.

The revised CKEditor is practical, but the UX is not on a par with Vizy. The link function is also not included in the core.

Don't get me wrong, Craft CMS is a very good CMS, but in terms of content input outside of static types, it definitely has many weaknesses. Also in the UX for content editing, some things would have to be optimized or at best plugins such as NEO, Vizy, Hyper and Co. would have to be adopted as first-party plugins.

Unfortunately, the web and content do not only consist of static fields.

When it comes to SEO, you can also build your own very good setup without SEOmatic with greater effort, but it takes time and you need many additional fields for this alone.

All in all, I'm a bit disappointed with Craft CMS 5, because we had hoped for more in the area of content editing and UX in content editing. It doesn't make Craft CMS worse, but it doesn't solve some of the problems it has had for many years.

And I can understand the price increase that has been made, because we have inflation, but a certain additional performance in the core would certainly be desirable. Since we are also speaking of software. Once something has been built here, we have completely different amortization rates than in other business areas.

We will not stop using Craft CMS, but we would have liked to see some changes.

**Reply by splendidrob** [reply]
@thupsi this is a really great solution thank you! I had been struggling to get my head around how to include multiple fields into one and of course the Matrix field allows this even if you don't want to repeat multiple items. This is a good way to keep everything native and have whatever options you wish across a button/cta/link style field.

**Reply by RyanRoberts** [reply]
@piotrpog Surely for some sites where SEOmatic is overkill, then SEOmate does the basics and more perfectly.

**Reply by mfell** [reply]
@splendidrob may I ask a question? (sorry for German screenshot) 
I tried to build this workaround: Craft 5, EntryType, MatrixField -> and put this in my ContentBuilder (NEO)

- I cannot create a new link (Attempt to read property "handle" on null) -> no idea what this means
- the UI it looks very confusing (for me ...)

Am I doing wrong? Is this the way of this workaround?

_PS: This topic is really frustating for me ... I rethink not to upgrade to Craft 5 and hope for a core solution in maybe one year.
And if not: Use Hyper. Feels not good for me. :/_

<img width="676" alt="Bildschirmfoto 2024-04-05 um 16 01 29" src="https://github.com/craftcms/cms/assets/6691773/71ca1f5f-e720-43c5-bfd4-dc2483c36f28">

**Reply by alexanderloewe** [reply]
@splendidrob: When you use neo, you could build one block, and use a dropdown field, and show/hide other fields via condition:

Neo Block:
– Dropdown (LinkType)
– Entry Field
– Asset Field
– URL Field
– E-Mail Field
– Phone Field
– Title Field
– Anchor Field
– Target Field
– Rel Field
...

Dropdown (LinkType):
– Entry Link
– Asset Link
– URL Link
– E-Mail Link
– Phone Link
...

Neo Block with condition:

Default: Entry Link (Dropdown)
– Show all fields for this type. (Entry Field, Title)
– Hide all other Fields

Asset Link (Dropdown)
– Show all fields for this type. (Asset Field, Title, Target)
– Hide all other Fields

...

BUT: This can make same issues with load performance.


––––––––

Or you simply make 5 blocks (one for each type)

**Reply by mfell** [reply]
@alexanderloewe Ok ... but then you have to build this structure within each Neo Block that contains a link. That's a lot in my usecase ... very hard to maintain.

Unfortunately it's not possible to build nested Neo Blocks.

**Reply by alexanderloewe** [reply]
@mfell

No, you do that once (per Builder)

Either you need a block with conditions
Or you build several BlockTyes in the 2nd variant.

You can then reactivate these as child blocks anywhere with a mouse click.

But none of the variants is a perfect solution

**Reply by mfell** [reply]
@alexanderloewe yepp ... your're right, thanks :)

But I agree ... none is a perfect solution.
At the moment, I think the best solution is just Hyper. (With a lot of stomach pain.)

**Reply by mfell** [reply]
@brandonkelly Dear Brandon, you are very busy with Craft 5.0.x I guess? May I ask if there is a chance for a core solution in Craft 5.x? If so, it would be nice to know for planning.

**Reply by thupsi** [reply]
@alexanderloewe While I agree that things can be improved, and you have every right to voice your concerns, it seems to me you are a bit harsh. The new features in Craft 5 are regarded by many, myself included, as huge steps forward in content modeling and editing. I, you, and others may think of a myriad ways to take this even further, but P&T have a finite amount of resources and they are pretty much developing non-stop. Heck, Brandon is doing commits every 5 minutes, even on weekends. Let's not forget that they are human beings, not robots 🤖. Burnout being a thing, especially in this industry, let's not contribute to it. Peace 😊

**Reply by alexanderloewe** [reply]
@thupsi: It's not a criticism of people, but rather of the priorities set for the Craft CMS project, which in the end will be carried by all developers - who choose this platform for themselves - and their customers.

I have been using Craft CMS for 9 years, and unfortunately some problems that were already known 9 years ago have not really been implemented as many developers would like. I haven't expressed my opinion here on Github for many years, but after talking to many developers who see similar things to me, I want to express it here.

I would like the P&T team to be more responsive to the wishes of the developers. Because at the end of the day, Craft CMS is one of the best systems for what we use it for every day. But it could be even better. If we were to draw more on concepts here and there that have already been tried and tested many times. Even if not all P&T developers like them.

All the more reason for me to hope that the P&T team is successful, because we enjoy using the product they build.

**Reply by cherrykoda** [reply]
Ran into this again today, so bumping the thread a bit.

**Reply by terryupton** [reply]
> Ran into this again today, so bumping the thread a bit. 

@cherrykoda Coming in 5.3 I think. 

https://github.com/craftcms/cms/pull/15251

**Reply by alexanderbuergin** [reply]
I hope!

**Comment by khalwat** [comment]
I agree that some kind of a link field (integrated with CKEditor) should be in core, primarily because it's a fundamental field type that is important to many Craft CMS builds.

There's already a URL field type baked in:

<img width="397" alt="Screenshot 2024-03-08 at 14 00 45" src="https://github.com/craftcms/cms/assets/7570798/1ac2d3cf-ffbd-4e92-9c0d-3cecf495ed27">

...but it's a little limited in scope, and doesn't allow linking to content in Craft.

Being able to create self-referencing links to content is so important to SEO and to UX for many sites, I feel that expanding the URL field to encompass the ability to link to arbitrary elements (and add links in Rich Text Fields) is important to the content authoring experience.

**Reply by mfell** [reply]
Yes ... but you mentioned it: It's too limited in scope / function.

**Comment by splendidrob** [comment]
Been exploring Craft 5 and options for this today and come up with a solution that for my needs at least works really well, just as good as the Link Field plugin and fully native. Many thanks to @thupsi and @stephanedemotte above who guided me in this direction! Gotta say I am loving how flexible the data structure is in Craft 5, its so much neater and lets you work more in a modular fashion within Craft itself similar to how we want to build out front end structures...Thanks so much for the update!

### 1: Create the entry type
I made an entry type of "Button" with the required fields:

- Drop down for linkType with options: none, url, entry and asset
- The title for the button text
- A toggle field for 'open in new tab?'
- The fields for a 'Url', 'Entry' and 'Asset' - these had conditionals applied so they will only show depending on the link type selected in the dropdown.
- A second drop down for optional styling (btn--primary btn--secondary etc)
- Some additional fields for colours (we have global colours for all buttons and now we can add optional overrides within the entry type, this is super useful!)

<img width="1266" alt="screenshot_1401" src="https://github.com/craftcms/cms/assets/25520763/8a1ad590-4340-48c0-970f-f8007e8936a9">

### 2: Create the field
I then created a matrix field "buttons" with a max of 2 so that we can output the "button" entry type wherever we want with a maximum of two buttons, generally a primary and secondary (I left this selectable with a dropdown but you could specify this in the code)
<img width="1060" alt="screenshot_1405" src="https://github.com/craftcms/cms/assets/25520763/954d7720-d1a7-46d1-84c2-3c266f2c1668">

We now have a super flexible "link" entry type that can be used either directly in entries or as a field inside other entry types or another matrix field.

Here are a couple of code examples for outputting (i've removed lots of other include settings for clarity) Built these quickly today so could probably be tidied up a fair bit:

**default.twig main page entry:**
```
{% include '_includes/cp.hero' with {
    buttons: entry.buttons,
  } %}
```

**cp.hero.twig include:**
```
{% set buttons = buttons ?? null %}

<section class="hero>
      {% if buttons | length %}
          {% for cta in buttons.all() %}
            {% include '_includes/cp.cta.twig' with {
              cta_type: cta.linkType,
              cta_text: cta.title,
              cta_ext_url: cta.externalLink,
              cta_entry: cta.entries.one(),
              cta_asset: cta.asset.one(),
              cta_newtab: cta.toggle,
              cta_classes: cta.buttonStyle == 'primary' ? 'btn w-100 w-auto@pw' : 'btn btn--secondary w-100 w-auto@pw',
              cta_bg_clr: cta.buttonBackgroundColour,
              cta_bg_hvr_clr: cta.buttonRolloverBackgroundColour,
              cta_txt_clr: cta.buttonTextColour,
              cta_txt_hvr_clr: cta.buttonRolloverTextColour,
            } %}
          {% endfor %}
      {% endif %}
</section><!--/hero-->
```

**cp.cta.twig our call to action or button include:**
```
{% set cta_type = cta_type | default('none') %}
{% set cta_text = cta_text | default('View' | t) %}
{% set cta_newtab = cta_newtab ?? null %}
{% set cta_classes = cta_classes | default('btn') %}

{% set cta_bg_clr = cta_bg_clr ?? null %}
{% set cta_bg_hvr_clr = cta_bg_hvr_clr ?? null %}
{% set cta_txt_clr = cta_txt_clr ?? null %}
{% set cta_txt_hvr_clr = cta_txt_hvr_clr ?? null %}

{% set cta_ext_url = cta_ext_url ?? null %}
{% set cta_entry = cta_entry ?? null %}
{% set cta_asset = cta_asset ?? null %}
{% set cta_url = cta_url ?? null %}

{% if cta_type == 'url' %}
  {% set cta_url = cta_ext_url %}
{% elseif cta_type == 'entry' %}
  {% set cta_url = cta_entry.url %}
{% elseif cta_type == 'asset' %}
  {% set cta_url = cta_asset.url %}
{% endif %}

{% if cta_url != 'none' %}
  <a href="{{ cta_url }}"
      {% if cta_newtab %}target="_blank"{% endif %}
      class="{{ cta_classes }}"
      style="
        --btn-bg-clr: {{ cta_bg_clr ? cta_bg_clr : branding.buttonBackgroundColour }};
        --btn-rollover-bg-clr: {{ cta_bg_hvr_clr ? cta_bg_hvr_clr : branding.buttonRolloverBackgroundColour }};
        --btn-clr: {{ cta_txt_clr ? cta_txt_clr : branding.buttonTextColour }};
        --btn-rollover-clr: {{ cta_txt_hvr_clr ? cta_txt_hvr_clr : branding.buttonRolloverTextColour }};
      "
  >
    {{ cta_text }}
  </a>
{% endif %}
```

Having gone through this process I can see why P&T don't have a super high priority for this as its definitely possible to build something way more flexible. Downsides obviously this is no use for old sites that already use the link field plugin! Currently we aren't able to update most of our old sites as nearly all link data is stored in the typed link field and this would require us going through and re-doing all the link content which is not viable. But for new sites this seems a good option moving forward and has enabled me to ditch a bunch of plugins - no need for supertable or neo etc anymore...

Hope this helps someone out and open to any suggestions as to how to improve!

**In use directly in an entry:** 
<img width="883" alt="screenshot_1402" src="https://github.com/craftcms/cms/assets/25520763/887f8190-0413-42d7-9379-69e2577e2a6c">

**In use inside a matrix field:**
<img width="1272" alt="screenshot_1404" src="https://github.com/craftcms/cms/assets/25520763/8b7c949c-85d7-4573-9ca9-a1f0778c9b38">

**Reply by CreateSean** [reply]
This is an interesting approach and I might adapt it, but I would still prefer a native field with options from P&T

**Reply by alexanderloewe** [reply]
On some projects, we use a similar setup. Well done!

**Reply by alexanderloewe** [reply]
@CreateSean We too ;)

**Reply by alexanderloewe** [reply]
I Instead of a link field, I would rather suggest something like an entry field, which does not have a type limit. So you could mix elements such as an entry, an asset or a product from Commerce in a «MultiType» field and output these contents in whatever form you want. This would solve more than one problem.

Yes, you can't map an external URL, an e-mail and a telephone kink with it, but everything else.

**Comment by brandonkelly** [comment, marked as answer]
Craft 5.3.0-beta.1 is out with a new Link field type 🎉 (#15251).

We’re targeting August 6 for the GA release. If you’d like to test ahead of time, change your `craftcms/cms` requirement in composer.json to:

```json
"craftcms/cms": "^5.3.0-beta.1",
```

and then run `composer update`.
